File: C++/045. Jump Game II.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n=nums.size();
        if(n<=1){return 0;}
        int lindex=0, hindex=0, minstep=0;
        while(lindex<=hindex)
        {
            int lasthindex=hindex;
            minstep++;
            for(int i=lindex;i<=lasthindex;i++)
            {
                int possibledist=i+nums[i];
                if(possibledist>=n-1){return minstep;}
                hindex=max(hindex,possibledist);
            }
            lindex=lasthindex+1;
        }
        return minstep;
    }
};

File: python/045. Jump Game II.py
class Solution(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        if n == 0: return -1
        end, nextEnd = 0, 0
        cnt = 0
        for i in range(n-1):
            nextEnd = max(nextEnd, i + nums[i])
            if i == end:
                if nextEnd == end: return -1
                end  = nextEnd
                cnt += 1
        if nextEnd >= n-1: return cnt
        else: return -1